Clean/Circuit/Theorems.lean

```diff
@@ -631,3 +631,120 @@ def FormalAssertion.isGeneralFormalCircuit (F : Type) (Input : TypeMap) [Field F
       rintro _ _ _ _ ⟨ _, _ ⟩
       apply orig.completeness <;> trivial
   }
+
+-- Structural rules for FormalCircuit
+
+namespace FormalCircuit
+
+/--
+Strengthen the assumptions of a FormalCircuit.
+Given a circuit with assumptions `Assumptions` and a stronger assumption `StrongerAssumptions`,
+this creates a new FormalCircuit with the stronger assumptions.
+-/
+def strengthenAssumptions {F : Type} [Field F] {Input Output : TypeMap} [ProvableType Input] [ProvableType Output]
+    (circuit : FormalCircuit F Input Output)
+    (StrongerAssumptions : Input F → Prop)
+    (h_stronger : ∀ input, StrongerAssumptions input → circuit.Assumptions input) :
+    FormalCircuit F Input Output where
+  elaborated := circuit.elaborated
+  Assumptions := StrongerAssumptions
+  Spec := circuit.Spec
+  soundness := fun offset env input_var input h_eval h_stronger_assumptions h_constraints =>
+    circuit.soundness offset env input_var input h_eval (h_stronger input h_stronger_assumptions) h_constraints
+  completeness := fun offset env input_var h_uses_local input h_eval h_stronger_assumptions =>
+    circuit.completeness offset env input_var h_uses_local input h_eval (h_stronger input h_stronger_assumptions)
+
+/--
+Derive a new specification from a FormalCircuit.
+Given a circuit and a proof that the original spec implies a new spec under the assumptions,
+this creates a new FormalCircuit with the derived specification.
+-/
+def deriveSpec {F : Type} [Field F] {Input Output : TypeMap} [ProvableType Input] [ProvableType Output]
```

Comment:
I think `weakenSpec` would be a more intuitive name here. Same for general formal circuit